songbird/src/driver/tasks/mixer.rs

```diff
@@ -0,0 +1,439 @@
+        // to volume, play state, etc.
+        let (mut len, mut opus_frame) =
+            self.mix_tracks(&mut opus_space, &mut mix_buffer, interconnect)?;
+
+        self.soft_clip.apply(&mut mix_buffer[..])?;
+
+        if self.muted {
+            len = 0;
+        }
+
+        if len == 0 {
+            if self.silence_frames > 0 {
+                self.silence_frames -= 1;
+
+                // Explicit "Silence" frame.
+                opus_frame = &SILENT_FRAME[..];
+            } else {
+                // Per official guidelines, send 5x silence BEFORE we stop speaking.
+                if let Some(ws) = &self.ws {
+                    // NOTE: this should prevent a catastrophic thread pileup.
+                    // A full reconnect might cause an inner closed connection.
+                    // It's safer to leave the central task to clean this up and
+                    // pass the mixer a new channel.
+                    let _ = ws.send(WsMessage::Speaking(false));
+                }
+
+                self.march_deadline();
+
+                return Ok(());
+            }
+        } else {
+            self.silence_frames = 5;
+        }
+
+        if let Some(ws) = &self.ws {
+            ws.send(WsMessage::Speaking(true))?;
+        }
+
+        self.march_deadline();
+        self.prep_and_send_packet(mix_buffer, opus_frame)?;
+
+        Ok(())
+    }
+
+    fn set_bitrate(&mut self, bitrate: Bitrate) -> Result<()> {
+        self.encoder.set_bitrate(bitrate).map_err(Into::into)
+    }
+
+    fn prep_and_send_packet(&mut self, buffer: [f32; 1920], opus_frame: &[u8]) -> Result<()> {
+        let conn = self
+            .conn_active
+            .as_mut()
+            .expect("Shouldn't be mixing packets without access to a cipher + UDP dest.");
+
+        let mut nonce = Nonce::default();
+        let index = {
+            let mut rtp = MutableRtpPacket::new(&mut self.packet[..]).expect(
+                "Too few bytes in self.packet for RTP header.\
+                    (Blame: VOICE_PACKET_MAX?)",
+            );
+
+            let pkt = rtp.packet();
+            let rtp_len = RtpPacket::minimum_packet_size();
+            nonce[..rtp_len].copy_from_slice(&pkt[..rtp_len]);
+
+            let payload = rtp.payload_mut();
+
+            let payload_len = if opus_frame.is_empty() {
+                self.encoder
+                    .encode_float(&buffer[..STEREO_FRAME_SIZE], &mut payload[TAG_SIZE..])?
+            } else {
+                let len = opus_frame.len();
+                payload[TAG_SIZE..TAG_SIZE + len].clone_from_slice(opus_frame);
+                len
+            };
+
+            let final_payload_size = TAG_SIZE + payload_len;
+
+            let tag = conn
+                .cipher
+                .encrypt_in_place_detached(&nonce, b"", &mut payload[TAG_SIZE..final_payload_size])
+                .expect("Encryption failed?");
```

Comment:
I think this would be better here, yes. I think I looked into when/where errors could appear from the XSalsa/**/ crate at the time we were migrating to the pure rust impl, and judged that encryption should never fail in this case, but I suppose it'll be better to guard against upstream changes here.